models/ResourceService.cfc

```diff
@@ -1,439 +1,517 @@
-
-			// Validate Location
-			if( NOT len( rbFullPath ) ){
-				throw("The resource bundle file: #rbFilePath# does not exist. Please check your path", "FullPath: #rbFullPath#", "ResourceBundle.InvalidBundlePath");
+			// Check unknown translation setting
+			if ( len( variables.unknownTranslation ) ) {
+				return variables.unknownTranslation & " key: #arguments.resource#";
 			}
 
-			//read file
-			fis = createObject("java","java.io.FileInputStream").init( rbFullPath );
-			rb = createObject("java","java.util.PropertyResourceBundle").init( fis );
-
-			try{
-				// Retrieve string
-				rbString = rb.handleGetObject( arguments.rbKey );
+			// Else return nasty unknown string.
+			return "_UNKNOWNTRANSLATION_FOR_#arguments.resource#_";
+		}
+
+		// Return Resource with value replacements
+		if ( structKeyExists( arguments, "values" ) ) {
+			return formatRBString( thisBundle[ arguments.resource ], arguments.values );
+		}
+
+		// return from bundle
+		return thisBundle[ arguments.resource ];
+	}
+	/************************************************************************************/
+	/****************************** UTILITY METHODS *************************************/
+	/************************************************************************************/
+
+	/**
+	 * Reads,parses and returns a resource bundle in struct format
+	 *
+	 * @rbFile This must be the path + filename UP to but NOT including the locale. We auto-add the local and .properties to the end.
+	 * @rbLocale The locale of the resource bundle
+	 */
+	struct function getResourceBundle( required rbFile, rbLocale = "en_US" ) {
+		var resourceBundle = {};
+		var thisKEY        = "";
+		var rbFilePath     = arguments.rbFile & iIf(
+			len( arguments.rbLocale ),
+			de( "_" ),
+			de( "" )
+		) & arguments.rbLocale & ".properties";
+		var rbFullPath = rbFilePath;
+
+		// Try to locate the path using the coldbox plugin utility
+		rbFullPath = variables.controller.locateFilePath( rbFilePath );
+
+		// Validate Location
+		if ( !len( rbFullPath ) ) {
+			throw(
+				"The resource bundle file: #rbFilePath# does not exist. Please check your path",
+				"FullPath: #rbFullPath#",
+				"ResourceBundle.InvalidBundlePath"
+			);
+		}
+
+		// create a file input stream with file location
+		var fis = createObject( "java", "java.io.FileInputStream" ).init( rbFullPath );
+		var fir = createObject( "java", "java.io.InputStreamReader" ).init( fis, "UTF-8" );
+		// Init RB with file Stream
+		var rb  = createObject( "java", "java.util.PropertyResourceBundle" ).init( fir );
+		try {
+			// Get Keys
+			var keys = rb.getKeys();
+
+			// Loop through property keys and store the values into bundle
+			while ( keys.hasMoreElements() ) {
+				thisKEY                   = keys.nextElement();
+				resourceBundle[ thisKEY ] = rb.handleGetObject( thisKEY );
 			}
-			catch(Any e){
-				fis.close();
-				$rethrow( e );
+		} catch ( Any e ) {
```

Comment:
Add a `final` block to close the file here instead.  We have this now.

```
finally {
   fis.close();
}

```

Reply:
done